app/assets/javascripts/discourse/app/mixins/composer-upload-uppy.js

```diff
@@ -343,6 +357,107 @@ export default Mixin.create({
     });
   },
 
+  _useS3MultipartUploads() {
+    const headers = {
+      "X-CSRF-Token": this.session.csrfToken,
+      "Content-Type": "application/json",
+    };
+    const self = this;
+
+    this._uppyInstance.use(AwsS3Multipart, {
+      // controls how many simultaneous _chunks_ are uploaded, not files,
+      // which in turn controls the minimum number of chunks presigned
+      // in each batch (limit / 2)
+      //
+      // the default, and minimum, chunk size is 5mb. we can control the
+      // chunk size via getChunkSize(file), so we may want to increase
+      // the chunk size for larger files
+      limit: 10,
+
+      createMultipartUpload(file) {
+        return ajax(getURL("/uploads/create-multipart.json"), {
+          type: "POST",
+          headers,
+          data: JSON.stringify({
+            file_name: file.name,
+            file_size: file.size,
+            content_type: file.type,
+            upload_type: file.meta.upload_type,
+          }),
+          // uppy is inconsistent, an error here fires the upload-error event
+        }).then((data) => {
+          file.meta.unique_identifier = data.unique_identifier;
+          return {
+            uploadId: data.external_upload_identifier,
+            key: data.key,
+          };
+        });
+      },
+
+      prepareUploadParts(file, partData) {
+        return (
+          ajax(getURL("/uploads/batch-presign-multipart-parts.json"), {
+            type: "POST",
+            headers,
+            data: JSON.stringify({
+              part_numbers: partData.partNumbers,
+              unique_identifier: file.meta.unique_identifier,
+            }),
+          })
+            .then((data) => {
+              return { presignedUrls: data.presigned_urls };
+            })
+            // uppy is inconsistent, an error here does not fire the upload-error event
+            .catch((err) => {
+              self._handleUploadError(file, err);
```

Comment:
Should this be the following instead?

```suggestion
              this._handleUploadError(file, err);
```

Reply:
No; see line 365. `this` here refers to the `prepareUploadsPart` function which is inside uppy, not the composer which is what we want